### PR TITLE
Occlusion sensitivity to use slidiing_window_inference

### PR DESCRIPTION
**Pull request by rijobro**
### Description

Occlusion sensitivity currently works by setting the logits at the center of the occluded region. Even though neighbouring voxels are occluded at the same time, this is not recorded. A better way to do it would be to do an average each time a voxel is occluded. This is taken care of by switching to using `sliding_window_inference`. 

We had to do a bit of hacking to get this to work. `sliding_window_inference` normally takes a subset of the whole image and infers that. We want `sliding_window_inference` to tell us which part of the image we should occlude, then we occlude it and infer the **whole** image. To that end, we use a meshgrid to tell us the coordinates of the region `sliding_window_inference` wants us to occlude. We occlude, infer and then crop back to the size of the region that was requested by `sliding_window_inference`. 

This PR also allows for different occlusion kernels. We currently have:
    - gaussian (actually inverse gaussian): the center of the occluded region is zero, and towards the edge of the image is unchanged. This doesn't introduce hard edges into the image, which might undermine the visualisation process.
    - mean_patch: the occluded region is replaced with the mean of the patch it is occluding.
    - mean_img: the occluded region is replaced with the mean of the whole image (current implementation).

## Changes to input arguments
This PR is backwards incompatible, as using `sliding_window_inference` means changing the API significantly.
    - pad_val: now determined by `mode`
    - stride: `overlap` used instead
    - per_channel: all channels are done simultaneously
    - upsampler: image is no longer downsampled

## Changes to output
Output previously had the shape `B,C,H,W,[D],N` where `C` and `N` were the number of input and output channels of the network, respectively. Now, we output the shape `B,N,H,W,[D]` as the `per_channel` feature is no longer present.

Columns 2-4 are occlusion sensitivity done with Gaussian, mean of patch and mean of image:

vis

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/visualize/occlusion_sensitivity.py [hunk 1]
```diff
@@ -142,139 +71,184 @@ class OcclusionSensitivity:
         - :py:class:`monai.visualize.occlusion_sensitivity.OcclusionSensitivity.`
     """
 
+    @deprecated_arg(
+        name="pad_val",
+        since="1.0",
+        removed="1.1",
```
Comment (wyli) [inline monai/visualize/occlusion_sensitivity.py:77]: we agreed to remove in the x+2 version, so I think this should be `1.2`: https://github.com/Project-MONAI/MONAI/blob/dev/CONTRIBUTING.md#backwards-compatibility
```suggestion
        removed="1.2",
```

File: monai/visualize/occlusion_sensitivity.py [hunk 2]
```diff
@@ -142,139 +71,184 @@ class OcclusionSensitivity:
...
+        kernel = torch.zeros((x.shape[1], *mask_size), device=x.device, dtype=x.dtype)
+        spatial_shape = kernel.shape[1:]
+        # all channels (as occluded shape already takes into account per_channel), center in spatial dimensions
+        center = [slice(None)] + [slice(s // 2, s // 2 + 1) for s in spatial_shape]
+        # place value of 1 at center
+        kernel[center] = 1.0
+        # Smooth with sigma equal to quarter of image, flip +ve/-ve so largest values are at edge
+        # and smallest at center. Scale to [0, 1].
+        gaussian = Compose(
+            [GaussianSmooth(sigma=[b // 4 for b in spatial_shape]), Lambda(lambda x: -x), ScaleIntensity()]
```
Comment (wyli) [inline monai/visualize/occlusion_sensitivity.py:151]: why choose this sigma as `b//4`? this might need some comments

Comment (rijobro) [reply monai/visualize/occlusion_sensitivity.py:151]: Empirically I found this to give good results. Below I've attached imshows of the kernels for mask sizes of 15 and 31. I don't think this parameter should be offered to the user to change, the goal is simply to give a smoothly varying kernel. What do you think? I can add to the documentation that it is empirically determined and give the code to do the imshow?

image

**Review by wyli — APPROVED** [review]
cool, please help address the minor comments inline...

looks like this sliding window implementation can close https://github.com/Project-MONAI/MONAI/issues/3032?

File: monai/visualize/occlusion_sensitivity.py (hunk 2 above)
Comment (wyli) [reply monai/visualize/occlusion_sensitivity.py:151]: thanks, I think your comment here is enough as a kind of documentation... so marking this as resolved.

Comment (rijobro) [reply monai/visualize/occlusion_sensitivity.py:151]: Few more plots, sigmas of 1, 0.25, 0.1 and mask sizes of 15 and 31. Think you'll agree that 0.25 looks best.
image

File: monai/visualize/occlusion_sensitivity.py [hunk 3]
```diff
@@ -301,29 +277,73 @@ def __call__(
...
             Both images will be cropped if a bounding box used, but voxel sizes will always match the input.
         """
+        if x.shape[0] > 1:
+            raise ValueError("Expected batch size of 1.")
+
+        sd = x.ndim - 2
+        mask_size = ensure_tuple_rep(self.mask_size, sd)
+
+        # get the meshgrid (so that sliding_window_inference can tell us which bit to occlude)
+        grid: MetaTensor = MetaTensor(
```
Comment (wyli) [inline monai/visualize/occlusion_sensitivity.py:287]: just another thought that by defining a binary mask to indicate the area to occlude, along with this `grid` variable, it seems it'll be possible to make occlusion of arbitrary shapes (not necessarily a set of rectangular bounding boxes).

**Review by ericspod — APPROVED** [review]

**Comment by ericspod** [comment]
This can close #3032 though one thing mentioned there is determining the importance of individual channels, this may be important to support with an example of how it can be shown to work. I'm not sure if cats and dogs would be suitable unless the network fixates on specific coat colours or something like that.

File: monai/visualize/occlusion_sensitivity.py [hunk 4]
```diff
@@ -142,139 +71,184 @@ class OcclusionSensitivity:
...
+        if callable(activate):
+            out = activate(out)
+        # else if True (should be boolean), sigmoid if n_chan == 1 else softmax
+        elif activate:
+            out = out.sigmoid() if x.shape[1] == 1 else out.softmax(1)
+
+        # the output will have shape [B,C] where C is number of channels output by model (inference classes)
+        # we need to return it to sliding window inference with shape [B,C,H,W,[D]], so add dims and repeat values
+        for m in mask_size:
+            out = torch.repeat_interleave(out.unsqueeze(-1), m, dim=-1)
```
[the remark this answers is not in the document]
Comment (rijobro) [reply monai/visualize/occlusion_sensitivity.py:227]: This will return the image as `BCH'W'D'`, where `H'W'D'` are the size of the mask, which is smaller than the image. If the device is big enough to handle `BCHWD`, then `BCH'W'D'` should be fine. The for loop is unfortunate but will only be run two or three times depending on the number of spatial dimensions in the image.

File: monai/visualize/occlusion_sensitivity.py (hunk 3 above)
Comment (rijobro) [reply monai/visualize/occlusion_sensitivity.py:287]: Yes! The Gaussian is already arguably circular/spherical as the voxels are affected radially from the centre of the occluded region.

**Comment by rijobro** [comment]
I don't think this solves #3032 – it still remains in the domain of interpretability of image classification, not segmentation. The output of `predictor` assumes the network gives the shape `BC`, not `BCHWD`.

**Comment by wyli** [comment]
> 

I see, for unet, it'll require some spatial pooling after this line to make it from BCHWD -> BC... 
https://github.com/Project-MONAI/MONAI/blob/e427a00babaa9c74cda7080b04a83e3a330799cd/monai/visualize/occlusion_sensitivity.py#L215
could be a very interesting future PR :)
